### DISCUSSION TITLE
JPEG compression of GTiff - artifacts around the edges

### DISCUSSION THREAD

**Discussion by mradamcox** [developmentseed/titiler#681] [General]
Hello!

I've been using TiTiler for a while on a project where I generate GeoTIFFs of historical maps and serve them to web apps. I've been experimenting with different compression methods for the files I make behind the scenes, and have come across a strange issue when I use JPEG compression.

In TiTiler, the edges of the warped image look like this:
![image](https://github.com/developmentseed/titiler/assets/10427268/16cde334-8184-4f29-aff6-efc146efc95a)
While the same edge in QGIS looks like
![image](https://github.com/developmentseed/titiler/assets/10427268/3feab43b-fd0b-4fab-8d27-a334d6fd26e6)

I have tried `nearest` and `average` resampling while creating the image with GDAL (and have tried all of the different resampling methods available in TiTiler for retrieving the tile) but the issue persists.

The file has an alpha band for a mask, and no overviews at the moment: https://loc-insurancemaps.us-southeast-1.linodeobjects.com/misc/augusta_wis_1910_p2__d2grZe_00-JPEG.tif

gdalinfo output:
```
Driver: GTiff/GeoTIFF
Files: ./augusta_wis_1910_p2__d2grZe_00-JPEG.tif
Size is 7888, 8935
Coordinate System is:
PROJCRS["WGS 84 / Pseudo-Mercator",
    BASEGEOGCRS["WGS 84",
        DATUM["World Geodetic System 1984",
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["Popular Visualisation Pseudo-Mercator",
        METHOD["Popular Visualisation Pseudo Mercator",
            ID["EPSG",1024]],
        PARAMETER["Latitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["easting (X)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["northing (Y)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["unknown"],
        AREA["World - 85°S to 85°N"],
        BBOX[-85.06,-180,85.06,180]],
    ID["EPSG",3857]]
Data axis to CRS axis mapping: 1,2
Origin = (-10143806.026370406150818,5571837.580980095081031)
Pixel Size = (0.073510744941599,-0.073510744941599)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  COMPRESSION=JPEG
  INTERLEAVE=PIXEL
  JPEGTABLESMODE=1
  JPEG_QUALITY=75
Corner Coordinates:
Upper Left  (-10143806.026, 5571837.581) ( 91d 7'24.10"W, 44d41' 0.73"N)
Lower Left  (-10143806.026, 5571180.762) ( 91d 7'24.10"W, 44d40'45.63"N)
Upper Right (-10143226.174, 5571837.581) ( 91d 7' 5.34"W, 44d41' 0.73"N)
Lower Right (-10143226.174, 5571180.762) ( 91d 7' 5.34"W, 44d40'45.63"N)
Center      (-10143516.100, 5571509.172) ( 91d 7'14.72"W, 44d40'53.18"N)
Band 1 Block=512x512 Type=Byte, ColorInterp=Red
  Mask Flags: PER_DATASET ALPHA 
Band 2 Block=512x512 Type=Byte, ColorInterp=Green
  Mask Flags: PER_DATASET ALPHA 
Band 3 Block=512x512 Type=Byte, ColorInterp=Blue
  Mask Flags: PER_DATASET ALPHA 
Band 4 Block=512x512 Type=Byte, ColorInterp=Alpha
```

Other methods of creating the file don't result in these artifacts in TiTiler (or QGIS):

- Compressed with DEFLATE: https://loc-insurancemaps.us-southeast-1.linodeobjects.com/misc/augusta_wis_1910_p2__jVr4zu_00-DEFLATE-PRED2.tif
- Used the `COG` driver in gdal.Translate with JPEG compression: https://loc-insurancemaps.us-southeast-1.linodeobjects.com/misc/augusta_wis_1910_p2__rf3eHN_00-COG.tif

One significant difference with the COG driver is that it doesn't create an alpha band, it (somehow) puts a mask into each of the RGB bands (there is no No Data value). Also the COG driver uses `YCbCr`, which is not possible with an alpha band. Perhaps this is the root of the artifact problem?

As things stand, I have some later operations that expect an alpha band in this image, but perhaps I should rework those later steps, so I can just use the COG driver here. Either way, I'd appreciate any thoughts on these edge artifacts, and whether they can be avoided in this scenario.

Thanks!

**Comment by vincentsarago** [comment]
@mradamcox 👋 you can't use ALPHA band with JPEG compression but only internal mask. I'm quite surprised that you were able to create an RGBA Jpeg compressed file 🤔 

TBH, I'm not quite sure to follow everything here...

**Comment by mradamcox** [comment]
Haha, yes sorry for the information overload. I have been deep in this GDAL rabbit hole...

I was kind of surprised I could make the file this way too, but with the right combination of warp / translate options it is possible. Everything seemed fine until I saw these edges.

However, sounds like this alpha band + JPEG compression is a bit of a hack, and it may be better to stick with more standard outputs.
